src/lazy/binary/system/lazy_struct.rs

```diff
@@ -0,0 +1,430 @@
+/// // Add up the integer values of all the fields named 'foo'
+/// let mut foo_sum = 0i64;
+/// for field in &lazy_struct {
+///     let field = field?;
+///     if field.name()? == "foo" {
+///         foo_sum += field.value().read()?.expect_i64()?;
+///     }
+/// }
+///
+/// assert_eq!(foo_sum, 4);
+///# Ok(())
+///# }
+/// ```
+pub struct LazyStruct<'top, 'data> {
+    pub(crate) raw_struct: LazyRawStruct<'data>,
+    pub(crate) symbol_table: &'top SymbolTable,
+}
+
+// Best-effort debug formatting for LazyStruct. Any failures that occur during reading will result
+// in the output being silently truncated.
+impl<'top, 'data> Debug for LazyStruct<'top, 'data> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{{")?;
+        for field in self {
+            let field = field.map_err(|_| fmt::Error)?;
+            let name = field.name().map_err(|_| fmt::Error)?;
+            let lazy_value = field.value();
+            let value = lazy_value.read().map_err(|_| fmt::Error)?;
+            write!(f, "{}:{:?},", name.text().unwrap_or("$0"), value)?;
+        }
+        write!(f, "}}")?;
+        Ok(())
+    }
+}
+
+impl<'top, 'data> LazyStruct<'top, 'data> {
+    /// Returns an iterator over this struct's fields. See [`LazyField`].
+    pub fn iter(&self) -> StructIterator<'top, 'data> {
+        StructIterator {
+            raw_struct_iter: self.raw_struct.iter(),
+            symbol_table: self.symbol_table,
+        }
+    }
+
+    /// Returns the value of the first field with the specified name, if any. The returned value is
+    /// a [`LazyValue`]. Its type and annotations can be inspected without calling [LazyValue::read].
+    ///
+    /// Because the `LazyStruct` does not store materialized values, it must seek over its fields
+    /// to find one with the requested name, giving this method linear time complexity.
+    /// ```
+    ///# use ion_rs::IonResult;
+    ///# fn main() -> IonResult<()> {
+    /// use ion_rs::element::Element;
+    /// use ion_rs::IonType;
+    /// use ion_rs::lazy::binary::lazy_reader::LazyReader;
+    /// use ion_rs::lazy::value_ref::ValueRef;
+    ///
+    /// let ion_data = r#"{foo: "hello", bar: quux::5, baz: null, bar: false}"#;
+    /// let ion_bytes: Vec<u8> = Element::read_one(ion_data)?.to_binary()?;
+    /// let mut reader = LazyReader::new(&ion_bytes)?;
+    ///
+    /// let lazy_struct = reader.expect_next()?.read()?.expect_struct()?;
+    ///
+    /// assert!(lazy_struct.find("foo")?.is_some());
+    /// assert!(lazy_struct.find("Ontario")?.is_none());
+    ///
+    /// // There are two 'bar' fields; `find` will return the value of the first.
+    /// let value = lazy_struct.find("bar")?.unwrap();
+    ///
+    /// assert!(value.annotations().next().unwrap()? == "quux");
+    /// assert_eq!(value.read()?, ValueRef::Int(5.into()));
+    ///
+    ///# Ok(())
+    ///# }
+    /// ```
+    pub fn find(&self, name: &str) -> IonResult<Option<LazyValue<'top, 'data>>> {
```

Comment:
> Should this (and things like it) be spelled with first (i.e., find_first)? Maybe that is too verbose, but I like being explicit about this in the API.

I'm torn. Explicitness is nice, but `find_first_expected` does feel a bit too heavyweight to me.

I think the doc comments are pretty clear about this, though:
```
    /// Returns the value of the first field with the specified name, if any. The returned value is
    /// a [`LazyValue`]. Its type and annotations can be inspected without calling [LazyValue::read].
    ///
    /// Because the `LazyStruct` does not store materialized values, it must seek over its fields
    /// to find one with the requested name, giving this method linear time complexity.
```

> Also, would it be mildly more ergonomic to have this be find<S: AsRef<str>>(&self, name: S) (work with String/Str and their references) or do you want this to be explicit?

I'm very open to adding this later. For this to really land there has to be a text reader/writer and some unifying traits, so I'm planning to wait until all of that shakes out first.